pkg/auth/providerrefresh/refresher_test.go

```diff
@@ -0,0 +1,192 @@
+					"local://user-abcde",
+				},
+			},
+			attribs: &v3.UserAttribute{
+				ObjectMeta:      metav1.ObjectMeta{Name: "user-abcde"},
+				GroupPrincipals: map[string]v3.Principals{},
+				ExtraByProvider: map[string]map[string][]string{},
+			},
+			tokens: []*v3.Token{
+				{
+					UserID:       "user-abcde",
+					IsDerived:    true,
+					AuthProvider: providers.LocalProvider,
+				},
+			},
+			want: &v3.UserAttribute{
+				ObjectMeta: metav1.ObjectMeta{Name: "user-abcde"},
+				GroupPrincipals: map[string]v3.Principals{
+					"local": v3.Principals{},
+				},
+				ExtraByProvider: map[string]map[string][]string{
+					providers.LocalProvider: map[string][]string{
+						common.UserAttributePrincipalID: []string{"local://user-abcde"},
+						common.UserAttributeUserName:    []string{"admin"},
+					},
+				},
+			},
+		},
+	}
+
+	providers.ProviderNames = map[string]bool{providers.LocalProvider: true}
+	providers.Providers = map[string]common.AuthProvider{
+		providers.LocalProvider: &mockLocalProvider{},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := &refresher{
+				tokenLister: &fakes.TokenListerMock{
+					ListFunc: func(_ string, _ labels.Selector) ([]*v3.Token, error) {
+						return tt.tokens, nil
+					},
+				},
+				userLister: &fakes.UserListerMock{
+					GetFunc: func(_, _ string) (*v3.User, error) {
+						return tt.user, nil
+					},
+				},
+			}
+			got, err := r.refreshAttributes(tt.attribs)
+			assert.Nil(t, err)
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+type mockLocalProvider struct {
+	userLister   managementv3.UserLister
+	groupLister  managementv3.GroupLister
+	userIndexer  cache.Indexer
+	gmIndexer    cache.Indexer
+	groupIndexer cache.Indexer
+	tokenMGR     *tokens.Manager
+	invalidHash  []byte
+}
+
+func (p *mockLocalProvider) GetName() string {
+	panic("not implemented")
+}
+
+func (p *mockLocalProvider) AuthenticateUser(ctx context.Context, input interface{}) (v3.Principal, []v3.Principal, string, error) {
+	panic("not implemented")
+}
+
+func (p *mockLocalProvider) SearchPrincipals(name, principalType string, myToken v3.Token) ([]v3.Principal, error) {
+	panic("not implemented")
+}
+
+func (p *mockLocalProvider) GetPrincipal(principalID string, token v3.Token) (v3.Principal, error) {
+	return v3.Principal{
+		ObjectMeta: metav1.ObjectMeta{Name: "admin"},
+		LoginName:  "admin",
+		Provider:   providers.LocalProvider,
+	}, nil
+}
+func (p *mockLocalProvider) CustomizeSchema(schema *types.Schema) {
+	panic("not implemented")
+}
+func (p *mockLocalProvider) TransformToAuthProvider(authConfig map[string]interface{}) (map[string]interface{}, error) {
+	panic("not implemented")
+}
+func (p *mockLocalProvider) RefetchGroupPrincipals(principalID string, secret string) ([]v3.Principal, error) {
+	return []v3.Principal{}, nil
+}
+func (p *mockLocalProvider) CanAccessWithGroupProviders(userPrincipalID string, groups []v3.Principal) (bool, error) {
+	return true, nil
```

Comment:
I think that if we keep the if statement in refresher.go as is, we want a test case or two where the user couldn't access the provider, so even though we had tokens we didn't use them and the user principals were empty.

Reply:
Added